macros/activity_schema/dataset/aggregations/first_value.sql

```diff
@@ -17,6 +17,19 @@ cast(split_part(
         ) as {{column.data_type}})
 {%- endmacro -%}
 
+{% macro bigquery__aggfunc_first_value(column) %}
```

Comment:
Nice! Only one functional note:
* It looks like the `min` function that wraps around the concatenated string is missing - check out line 10 in this file for reference. (That's probably what's causing the `group by` error in your project.) Same goes for the `max` function needed in the `last_value` macro

On a separate note note - looking at the code, I'm undecided on whether or not this macro should have an adapter-specific implementation for a couple reasons - first, the only difference from the `default` is the use of `dbt.split_part` instead of native `split_part`, and `dbt.split_part` works for all other warehouses as well. Second, I sometimes forget how ugly it gets to have to convert a sql snippet to a string in order to use it as a macro argument 😓. I could also see an argument in favor of splitting out so that the code change is guaranteed not to break any existing usage.

Let me know what you think!

Reply:
Ah yes. The group by issue makes sense when you completely miss the min/max which is the whole point of the macro. 😞 

I'll make that change in the morning. 

Yeah my thought was to not break any other possible use for a different adapter other than BigQuery. Your call, but I'd argue for making BigQuery separate as you've already done on other macros - but my opinion is biased 😄 .

Reply:
I'm on board with the separate implementation! One aesthetic ask, though - for the query snippet that you pass as the first argument in `dbt.split_part`, would you mind making it a bit more readable? One option is to define it as a variable in a jinja set block and then pass the variable to the macro - but I'm sure there are plenty of other reasonable alternatives as well.

Reply:
I made the changes requested, but now I'm running into a json-related error: 
`The provided JSON input is not a string`

This occurs in both the primary activity cte - for json data types of int() but not the types of string()
I can comment those columns out for now and successfully run a query to select from the primary cte, however, the issue arises in the secondary activity cte and occurs on both json types of int() as expected, but also on types of string().

I think a more appropriate fix is to not require so many cast statements and to try and simplify how the json fields are parsed out for use. But not clear to me yet where to go with this next.

Reply:
Ah I see - your intuition makes sense around the excess data type casting. Can you share the compiled sql for one of the queries that failed? Ideally one that would throw an error for both the primary and joined activity.

Reply:
See below:
```

  
    

    create or replace table `project.dataset.table`
      
    
    

    OPTIONS(
      expiration_timestamp=TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 168 hour),
    
      description=""""""
    )
    as (
      



-- depends_on: `project.dataset`.`audience`



    with

primary__all__email_recipient_entered__None__1 as (
    select
        -- required columns to use for joins
        p.activity_id as req__activity_id,
        p.email as req__email,
        p.ts as req__ts,
        p.activity_occurrence as req__activity_occurrence,
        p.activity_repeated_at as req__activity_repeated_at,
        p.ts as email_recipient_entered_at,
        p.email as email_recipient_entered_select_all__email,
        cast(nullif(string(json_extract(p.attributes, '$.date_picked_source')), '') as STRING) as email_recipient_entered_select_all__date_picked_source,
        cast(nullif(string(json_extract(p.attributes, '$.domain')), '') as STRING) as email_recipient_entered_select_all__domain,
        cast(nullif(string(json_extract(p.attributes, '$.job_title')), '') as STRING) as email_recipient_entered_select_all__job_title,
        cast(nullif(string(json_extract(p.attributes, '$.status')), '') as STRING) as email_recipient_entered_select_all__status
    from `project.dataset`.`audience` as p
    where p.activity = 'email recipient entered'
        and true
),


joined__append__first__after__clicked_email_link__None__1 as (
    
    select
        p.req__activity_id,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(j.ts as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as TIMESTAMP) as clicked_email_link_at,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(cast(nullif(string(json_extract(j.attributes, '$.brand_id')), '') as INT64) as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as INT64) as clicked_email_link_append_first__after_brand_id,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(cast(nullif(string(json_extract(j.attributes, '$.brand_title')), '') as STRING) as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as STRING) as clicked_email_link_append_first__after_brand_title,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(cast(nullif(string(json_extract(j.attributes, '$.campaign_id')), '') as INT64) as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as INT64) as clicked_email_link_append_first__after_campaign_id,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(cast(nullif(string(json_extract(j.attributes, '$.campaign_title')), '') as STRING) as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as STRING) as clicked_email_link_append_first__after_campaign_title,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(cast(nullif(string(json_extract(j.attributes, '$.mailing_id')), '') as INT64) as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as INT64) as clicked_email_link_append_first__after_mailing_id,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(cast(nullif(string(json_extract(j.attributes, '$.mailing_title')), '') as STRING) as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as STRING) as clicked_email_link_append_first__after_mailing_title,
        cast(
    

  
    split(
        min(cast(j.ts as STRING) || ';.,;' || cast(cast(nullif(string(json_extract(j.attributes, '$.url')), '') as STRING) as STRING)),
        ';.,;'
        )[safe_offset(1)]
  


as STRING) as clicked_email_link_append_first__after_url
    from primary__all__email_recipient_entered__None__1 as p
    left join `project.dataset.`audience` j
        -- filter joined activity first to improve query performance
        on j.activity = 'clicked email link'
        and p.req__email = j.email
        and p.req__ts < j.ts
        
    group by 1
    
)

select
    p.email_recipient_entered_at,
    p.email_recipient_entered_select_all__email,
    p.email_recipient_entered_select_all__date_picked_source,
    p.email_recipient_entered_select_all__domain,
    p.email_recipient_entered_select_all__job_title,
    p.email_recipient_entered_select_all__status,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_at as clicked_email_link_at,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_append_first__after_brand_id as clicked_email_link_append_first__after_brand_id,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_append_first__after_brand_title as clicked_email_link_append_first__after_brand_title,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_append_first__after_campaign_id as clicked_email_link_append_first__after_campaign_id,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_append_first__after_campaign_title as clicked_email_link_append_first__after_campaign_title,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_append_first__after_mailing_id as clicked_email_link_append_first__after_mailing_id,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_append_first__after_mailing_title as clicked_email_link_append_first__after_mailing_title,
    joined__append__first__after__clicked_email_link__None__1.clicked_email_link_append_first__after_url as clicked_email_link_append_first__after_url
from primary__all__email_recipient_entered__None__1 as p
left join joined__append__first__after__clicked_email_link__None__1
    on p.req__activity_id = joined__append__first__after__clicked_email_link__None__1.req__activity_id


    );
 ```